### PR TITLE
implement ST_SharedPaths

### PR DESCRIPTION
**Pull request by devenbhooshan**
Implemented the geometry based builtin `ST_SharedPaths`

### CODE REVIEW

**Comment by devenbhooshan** [comment]
@otan please review.

File: pkg/geo/geomfn/shared_paths_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			name: "shared path between a linestring and multilinestring",
+			args: args{
+				a: geo.MustParseGeometry("LINESTRING(76 175,90 161,126 125,126 156.25,151 100)"),
+				b: geo.MustParseGeometry("MULTILINESTRING((26 125,26 200,126 200,126 125,26 125), " +
+					"(51 150,101 150,76 175,51 150))"),
+			},
+			want: geo.MustParseGeometry("GEOMETRYCOLLECTION(MULTILINESTRING EMPTY," +
+				"MULTILINESTRING((76 175,90 161),(90 161,101 150),(126 125,126 156.25)))"),
+		},
```
Comment (otan) [inline pkg/geo/geomfn/shared_paths_test.go:50]: can you add a case for no shared paths here?

File: pkg/geo/geomfn/shared_paths_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		b *geo.Geometry
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *geo.Geometry
+		wantErr bool
+	}{
+		{
+			name: "shared path between a multilinestring and linestring",
```
Comment (otan) [inline pkg/geo/geomfn/shared_paths_test.go:32]: can you capitalise these as LineString / MultiLineString?

File: pkg/geo/geomfn/shared_paths_test.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			args: args{
+				a: geo.MustParseGeometry("LINESTRING(76 175,90 161,126 125,126 156.25,151 100)"),
+				b: geo.MustParseGeometry("MULTILINESTRING((26 125,26 200,126 200,126 125,26 125), " +
+					"(51 150,101 150,76 175,51 150))"),
+			},
+			want: geo.MustParseGeometry("GEOMETRYCOLLECTION(MULTILINESTRING EMPTY," +
+				"MULTILINESTRING((76 175,90 161),(90 161,101 150),(126 125,126 156.25)))"),
+		},
+	}
+	for _, tt := range tests {
```
Comment (otan) [inline pkg/geo/geomfn/shared_paths_test.go:52]: can you add tests for points?

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 4]
```diff
@@ -3825,6 +3825,35 @@ Bottom Left.`,
...
+	"st_sharedpaths": makeBuiltin(
+		defProps(),
+		geometryOverload2(
+			func(ctx *tree.EvalContext, a, b *tree.DGeometry) (tree.Datum, error) {
+				ret, err := geomfn.SharedPaths(a.Geometry, b.Geometry)
+				if err != nil {
+					return nil, err
+				}
+
+				if ret == nil {
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:3840]: is this correct? what cases is this correct?

File: pkg/sql/logictest/testdata/logic_test/geospatial [hunk 5]
```diff
@@ -4285,3 +4285,12 @@ POINT (10.5 20.25)
 
 statement error unknown function: public.log\(\), but log\(\) exists
 SELECT public.log(10)
+
+query T
+SELECT ST_AsText(
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:4290]: can you place this near the `geom_operators_tests`?

**Review by otan — COMMENTED** [review]
Hey -- the approach look ok, just a few more test cases and details I'd like for you to work out.

Also, please follow blathers' suggestions.

File: pkg/geo/geomfn/shared_paths.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
```
Comment (otan) [inline pkg/geo/geomfn/shared_paths.go:13]: mind moving this to `topology_operations.go` as well?

Comment (devenbhooshan) [reply pkg/geo/geomfn/shared_paths.go:13]: sharedpaths is[ not a topological operation](http://www.gitta.info/SpatialQueries/en/html/TopoBasedOps_learningObject2.html#:~:text=Topological%20operators%20are%20components%20of,as%20ArcGIS%2C%20Geomedia%20or%20MapInfo.&text=Further%20topological%20operators%2C%20which%20are,corresponding%20data%20structure%2C%20are%20developed.) right ?

File: pkg/geo/geomfn/shared_paths_test.go (hunk 1 above)
Comment (devenbhooshan) [reply pkg/geo/geomfn/shared_paths_test.go:51]: working on this. facing an issue with empty geomtrycollection. the actual and the expected length of the EWKB differs. checking this.. 

```
        shared_paths_test.go:77:
            	Error Trace:	shared_paths_test.go:77
            	Error:      	Not equal:
            	            	expected: &geo.Geometry{spatialObject:geopb.SpatialObject{Type:2, EWKB:geopb.EWKB{0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, SRID:0, ShapeType:8, BoundingBox:(*geopb.BoundingBox)(nil)}}
            	            	actual  : &geo.Geometry{spatialObject:geopb.SpatialObject{Type:2, EWKB:geopb.EWKB{0x1, 0x7, 0x0, 0x0, 0x0, 0x2, 0x0, 0x0, 0x0, 0x1, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, SRID:0, ShapeType:8, BoundingBox:(*geopb.BoundingBox)(nil)}}

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,5 @@
            	            	   Type: (geopb.SpatialObjectType) 2,
            	            	-  EWKB: (geopb.EWKB) (len=9) {
            	            	-   00000000  01 07 00 00 00 00 00 00  00                       |.........|
            	            	+  EWKB: (geopb.EWKB) (len=27) {
            	            	+   00000000  01 07 00 00 00 02 00 00  00 01 05 00 00 00 00 00  |................|
            	            	+   00000010  00 00 01 05 00 00 00 00  00 00 00                 |...........|
            	            	   },
```

File: pkg/geo/geomfn/shared_paths.go (hunk 6 above)
Comment (otan) [reply pkg/geo/geomfn/shared_paths.go:13]: it's how it's grouped in GEOS, heh!

https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in#L544

File: pkg/geo/geomfn/shared_paths_test.go (hunk 1 above)
Comment (otan) [reply pkg/geo/geomfn/shared_paths_test.go:51]: looks like
```

otan=# select st_astext('010700000002000000010500000000000000010500000000000000'::geometry);
                            st_astext
-----------------------------------------------------------------
 GEOMETRYCOLLECTION(MULTILINESTRING EMPTY,MULTILINESTRING EMPTY)
(1 row)

```

**Comment by devenbhooshan** [comment]
@otan  incorporated the comments. please check. 

one thing I wanted to highlight. `pkg/sql/sem/builtins/geo_builtins.go` is huge. we should try to decompose it into smaller files. same issue with `pkg/sql/logictest/testdata/logic_test/geospatial`

**Review by otan — APPROVED** [review]

**Comment by otan** [comment]
> one thing I wanted to highlight. pkg/sql/sem/builtins/geo_builtins.go is huge. we should try to decompose it into smaller files. same issue with pkg/sql/logictest/testdata/logic_test/geospatial

yeah it's kind of crept up. haven't had the time to work on this yet against other priorities :( - but we have an upcoming stability period to get on it.

**Comment by devenbhooshan** [comment]
thanks, @otan for giving the green signal. 

@otan would love to help with the refactoring.

**Comment by devenbhooshan** [comment]
```
panic: only 2D objects are currently supported [recovered]
```
from the logs.

**Comment by otan** [comment]
yeah i've rebased that.
the recent failure is a flake.

bors r+

**Comment by SWDevAngel** [comment]
@devenbhooshan LAST DAY TO ORDER!
Hi Deven. Thank you for contributing to CockroachDB this year. As a token of our appreciation we would like to send you a gift. We tried to reach you on email, but if you didn’t see the message please DM me on our [community Slack](https://cockroa.ch/slack) `@lisa` so I can send you a link to order your gift. (If you don’t want a gift, we also have a charitable contribution choice.) All orders must be in by December 13 (that’s tomorrow!), so please send this asap. :)

**Pull request merged**